great_expectations/data_context/store/datasource_store.py

```diff
@@ -172,22 +167,31 @@ def set_by_name(
         )
         self.set(datasource_key, datasource_config)
 
-    def create(
-        self, datasource_config: DatasourceConfig
-    ) -> Optional[GeCloudResourceRef]:
+    def set(  # type: ignore[override]
+        self, key: Union[DataContextKey, None], value: DatasourceConfig, **_: dict
+    ) -> DatasourceConfig:
         """Create a datasource config in the store using a store_backend-specific key.
-
         Args:
-            datasource_config: Config containing the datasource name.
-
+            key: Optional key to use when setting value.
+            value: DatasourceConfig set in the store at the key provided or created from the DatasourceConfig attributes.
+            **_: kwargs will be ignored but accepted to align with the parent class.
         Returns:
-            None unless using GeCloudStoreBackend and if so the GeCloudResourceRef which contains the id
-            which was used to create the config in the backend.
+            DatasourceConfig retrieved from the DatasourceStore.
         """
-        key: Union[
-            GeCloudIdentifier, DataContextVariableKey
-        ] = self._build_key_from_config(datasource_config)
-        return self.set(key, datasource_config)  # type: ignore[func-returns-value]
+        if not key:
+            key = self._build_key_from_config(value)
+
+        # Make two separate requests to set and get in order to obtain any additional
+        # values that may have been added to the config by the StoreBackend (i.e. object ids)
+        ref: Optional[Union[bool, GeCloudResourceRef]] = super().set(key, value)  # type: ignore[func-returns-value]
+        if ref and isinstance(ref, GeCloudResourceRef):
+            key.ge_cloud_id = ref.ge_cloud_id  # type: ignore[attr-defined]
```

Comment:
We need to use the id to retrieve the datasource config we just set in the store. We can consider other ways to do this, e.g. check if ref is GeCloudResourceRef and then set `key = ref`

Reply:
This is an unfortunate compromise that leaks; I'd rather have the set method responsible for returning a reference to whatever resource was just created/updated _for all store backends_ and remove `bool` as an expected return value. Maybe this one piece can be captured in a JIRA item? I understand that we had to do it this way for now

Reply:
Just replied above about this. I agree about capturing in a separate JIRA and about not being excited about this approach. Created JIRA: https://superconductive.atlassian.net/browse/GREAT-1235 - is there a good epic to put this in?